### PR TITLE
Preparation work to add a security margin in collision tests

### PR DESCRIPTION
**Pull request by florent-lamiraux**
- A lot of cleaning among functions that were not used
    - broadphase has been removed,
    - cost and cost source have been removed,

  - CollisionRequest is passed to low level functions in order to unify
    the calling function tree among the various possible options,

  - Collision test triangle - triangle
    - in GJK, a detection of loop that made the algorithm fail to detect
      collision has been removed,
    - collision and distance computation between triangles in
      MeshCollisionTraversalNode::leafTesting now uses GJK
    - a test on triangle - triangle interaction has been added optimality of
      closest points is check using Karush-Kuhn-Tucker conditions.

### CODE REVIEW

File: include/hpp/fcl/BV/AABB.h [hunk 1]
```diff
@@ -95,9 +95,11 @@ class AABB
   }    
 
   /// Not implemented
-  inline bool overlap(const AABB&, FCL_REAL&) const
+  inline bool overlap(const AABB& other, const CollisionRequest&,
+                      FCL_REAL& sqrDistLowerBound) const
   {
-    throw std::runtime_error ("Not implemented");
+    sqrDistLowerBound = sqrt (-1);
+    return overlap (other);
```
Comment (jmirabel) [inline include/hpp/fcl/BV/AABB.h]: Overlapping between AABB with distance lower bound should be easy to implement. Something like:
```cpp
if((min_.array() > other.max_.array()).any()) {
  sqrDistLowerBound = (min_ - other.max_).cwiseMax(0).squaredNorm();
  return false;
}
if((max_.array() < other.min_.array()).any()) {
  sqrDistLowerBound = (max_ - other.min_).cwiseMin(0).squaredNorm();
  return false;
}
sqrDistLowerBound = 0;
return true;
```

File: include/hpp/fcl/traversal/traversal_node_octree.h [hunk 2]
```diff
@@ -347,50 +346,19 @@ class OcTreeSolver
...
         return false;
     }
 
     /// stop when 1) bounding boxes of two objects not overlap; OR
     ///           2) at least of one the nodes is free; OR
     ///           2) (two uncertain nodes or one node occupied and one node uncertain) AND cost not required
-    if(tree1->isNodeFree(root1) || s.isFree()) return false;
-    else if((tree1->isNodeUncertain(root1) || s.isUncertain()) && !crequest->enable_cost) return false;
+    if(tree1->isNodeFree(root1)) return false;
+    else if((tree1->isNodeUncertain(root1) || s.isUncertain())) return false;
```
Comment (jmirabel) [inline include/hpp/fcl/traversal/traversal_node_octree.h]: Don't we want to remove this and only do the code in `else` part below ?

File: include/hpp/fcl/traversal/traversal_node_octree.h [hunk 3]
```diff
@@ -600,60 +559,21 @@ class OcTreeSolver
...
     }
 
     /// stop when 1) bounding boxes of two objects not overlap; OR
     ///           2) at least one of the nodes is free; OR
     ///           2) (two uncertain nodes OR one node occupied and one node uncertain) AND cost not required
-    if(tree1->isNodeFree(root1) || tree2->isFree()) return false;
-    else if((tree1->isNodeUncertain(root1) || tree2->isUncertain()) && !crequest->enable_cost) return false;
+    if(tree1->isNodeFree(root1)) return false;
+    else if((tree1->isNodeUncertain(root1) || tree2->isUncertain()))
+      return false;
```
Comment (jmirabel) [inline include/hpp/fcl/traversal/traversal_node_octree.h]: Don't we want to remove this and only do the code in `else` part below ?

File: include/hpp/fcl/traversal/traversal_node_octree.h [hunk 4]
```diff
@@ -952,7 +825,8 @@ class OcTreeSolver
     ///           2) at least one of the nodes is free; OR
     ///           2) (two uncertain nodes OR one node occupied and one node uncertain) AND cost not required
     if(tree1->isNodeFree(root1) || tree2->isNodeFree(root2)) return false;
-    else if((tree1->isNodeUncertain(root1) || tree2->isNodeUncertain(root2)) && !crequest->enable_cost) return false;
+    else if((tree1->isNodeUncertain(root1) || tree2->isNodeUncertain(root2)))
+      return false;
```
Comment (jmirabel) [inline include/hpp/fcl/traversal/traversal_node_octree.h]: Don't we want to remove this and only do the code in `else` part below ?

File: src/BV/OBB.cpp [hunk 5]
```diff
@@ -294,84 +295,44 @@ bool obbDisjoint(const Matrix3f& B, const Vec3f& T, const Vec3f& a, const Vec3f&
...
-
-  diff = t - (b[2] + Bf.col(2).dot(a));
-  if (diff > 0) {
-    squaredLowerBoundDistance += diff*diff;
-  }
+  AABB_corner = (B.transpose () * T).cwiseAbs ()  - Bf.transpose () * a;
+  // diff3 = | B^T T| - b - Bf^T a
+  diff3 = AABB_corner - b;
+  diff3 = diff3.cwiseMax (0);
+  squaredLowerBoundDistance = diff3.squaredNorm ();
```
Comment (jmirabel) [inline src/BV/OBB.cpp]: `squaredLowerBoundDistance = (AABB_corner - b).cwiseMax(0).squaredNorm()` might be better optimized by the compiler (though I am not sure).

File: include/hpp/fcl/BV/AABB.h (hunk 1 above)
Comment (florent-lamiraux) [reply include/hpp/fcl/BV/AABB.h]: Probably, but as we do not use this part of the code, I prefer not implement it without testing it.

File: include/hpp/fcl/collision_object.h [hunk 6]
```diff
@@ -88,13 +86,13 @@ class CollisionGeometry
   }
 
   /// @brief whether the object is completely occupied
-  inline bool isOccupied() const { return cost_density >= threshold_occupied; }
+  inline bool isOccupied() const { return true; }
```
[the remark this answers is not in the document]
Comment (florent-lamiraux) [reply include/hpp/fcl/collision_object.h]: You are right.

File: include/hpp/fcl/traversal/traversal_node_bvhs.h [hunk 7]
```diff
@@ -55,14 +57,44 @@
...
+     const Transform3f& tf1, const Transform3f& tf2, Vec3f& normal)
+    {
+      Vec3f globalP1 (tf1.transform (P1));
+      Vec3f globalP2 (tf1.transform (P2));
+      Vec3f globalP3 (tf1.transform (P3));
+      Vec3f globalQ1 (tf2.transform (Q1));
+      Vec3f globalQ2 (tf2.transform (Q2));
+      Vec3f globalQ3 (tf2.transform (Q3));
+      Vec3f u ((globalP2-globalP1).cross (globalP3-globalP1));
+      normal = u.normalized ();
```
[the remark this answers is not in the document]
Comment (florent-lamiraux) [reply include/hpp/fcl/traversal/traversal_node_bvhs.h]: I would rather filter degenerated triangles when building the model.

File: include/hpp/fcl/traversal/traversal_node_octree.h (hunk 2 above)
Comment (florent-lamiraux) [reply include/hpp/fcl/traversal/traversal_node_octree.h]: I did not change anything in the octree part where the notion of free and occupied is probably useful.

File: include/hpp/fcl/traversal/traversal_node_octree.h (hunk 3 above)
Comment (florent-lamiraux) [reply include/hpp/fcl/traversal/traversal_node_octree.h]: Same comment as above.

**Comment by jmirabel** [comment]
I fine with those changes. I let you merge when ready.

**Comment by jmirabel** [comment]
I fine with those changes. I let you merge when ready.

**Pull request merged**